src/instruction.rs

```diff
@@ -1080,19 +1131,91 @@ mod test_instruction_display {
 #[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
 pub enum Qubit {
     Fixed(u64),
+    #[serde(skip)] // todo
+    Placeholder(QubitPlaceholder),
     Variable(String),
 }
 
+impl Qubit {
+    fn resolve_placeholder<R>(&mut self, resolver: R)
+    where
+        R: Fn(&QubitPlaceholder) -> Option<u64>,
+    {
+        if let Qubit::Placeholder(placeholder) = self {
+            if let Some(resolved) = resolver(placeholder) {
+                *self = Qubit::Fixed(resolved);
+            }
+        }
+    }
+
+    // For comparison, this takes a more active approach than the above, in that it can allocate new indices
+    // using the placeholder. This seems unnecessary and so I've commented it out, but leaving it here
+    // for commentary while in draft
+    // fn resolve_placeholder<I, R>(
+    //     &mut self,
+    //     indices: &mut I,
+    //     resolver: R,
+    // ) -> Result<Option<(QubitPlaceholder, u64)>, ()>
+    // where
+    //     I: Iterator<Item = u64>,
+    //     R: Fn(QubitPlaceholder) -> Option<u64>,
+    // {
+    //     if let Qubit::Placeholder(placeholder) = self {
+    //         if let Some(resolved) = resolver(placeholder.clone()) {
+    //             *self = Qubit::Fixed(resolved);
+    //             Ok(None)
+    //         } else if let Some(next) = indices.next() {
+    //             let return_value = Some((placeholder.clone(), next));
+    //             *self = Qubit::Fixed(next);
+    //             Ok(return_value)
+    //         } else {
+    //             Err(())
+    //         }
+    //     } else {
+    //         Ok(None)
+    //     }
+    // }
+}
+
 impl fmt::Display for Qubit {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         use Qubit::*;
         match self {
             Fixed(value) => write!(f, "{}", value),
+            Placeholder(_) => todo!(),
             Variable(value) => write!(f, "{}", value),
         }
     }
 }
 
+type QubitPlaceholderInner = Arc<Pin<Box<usize>>>;
+
+/// An opaque placeholder for a qubit whose index may be assigned
+/// at a later time.
+#[derive(Clone, Debug, Eq)]
+pub struct QubitPlaceholder(QubitPlaceholderInner);
+
+impl Default for QubitPlaceholder {
+    fn default() -> Self {
+        use rand::Rng;
+        let value = rand::thread_rng().gen();
```

Comment:
It's opaque so the details shouldn't matter too much, and I doubt we'd ever create enough of these to worry about collisions, but: do we need these to be pseudorandom? Why not a UUID, or even just increment a counter?

```rust
use std::sync::atomic::{AtomicUsize, Ordering};

impl Default for QubitPlaceholder {
    fn default() -> Self {
        static COUNTER: AtomicUsize = AtomicUsize::default();
        let value = COUNTER.fetch_add(1, Ordering::Relaxed);
        Self(Arc::new(Box::pin(value)))
    }
}
```

Reply:
Yep, they don't. The thought here was to support serde serialization but now that that's been ruled out, it's not needed. The inner type is now `()`, so we pretty much just use the Arc for its own address directly.

Reply:
As far as counter, something would need to store that counter's state, and there's no eligible container as far as we could determine - _maybe_ a static var but that feels messy. Couldn't be a program since a placeholder should be valid and unique across programs or out of the scope of a program entirely

Reply:
Fortunately this has been obviated altogether